### ISSUE TITLE
VNET Integration for Azure function using Consumption plan

### ISSUE THREAD

**Issue by MayankSri** [Azure/Azure-Functions#646]
My azure function is running on a Consumption plan and it needs to access a resource running on a VM on a Azure VNET. The resource cannot be exposed via http.

Is there a solution other then switching to App Service Plan?

**Comment by pragnagopa** [comment]
Consumption Plan does not have support VNET.

**Comment by mnns** [comment]
Are there any plans for VNET for consumption plan? Serverless meaning = not handling the servers. It seems the true "serverless" Azure Functions (consumption) has little to no features such as the basic Virtual Network (simple basic isolation). Going to App Service Plan means it loses the "serverless" idea.

**Comment by pragnagopa** [comment]
Reactivating the issue.

**Comment by pragnagopa** [comment]
@btardif / @ColbyTresness  - Do we have any plans for adding support of VNET in consumption plan?

**Comment by ColbyTresness** [comment]
There are some technical blockers that make this really challenging, but it something we're aware of as an ask and are tracking. No concrete plans, however.

**Comment by nzthiago** [comment]
As discussed with @jeffhollan we have had this same request from other high profile customers.

More info:
They want to utilize Azure Functions for replacing an antiquated, hard to update applications while making way into other possible applications that can utilize Azure.
They have a lot of resources inside a VNET and need to be able to reach the different endpoints hosted there.
They prefer and their workload is conducive to the consumption plan. 
So they ask for VNET support on Consumption Plan for Azure Functions.

This would unblock other workloads for them besides this initial project.

**Comment by mrtristan** [comment]
as someone that's fully on AWS yet starting to look around at leveraging other tools like azure's functions, it seems that this missing capability is barring me from being able to work securely with my existing infrastructure.

**Comment by alexkarcher-msft** [comment]
We are working on adding this functionality to a serverless SKU. I'll have more info in the coming months :)

**Comment by AlainPilon** [comment]
+1 This is a key feature, and blocker for a few things I am building as the consumption plan is the only way to make a lot of small initiative financially viable. 👀

**Comment by alexkarcher-msft** [comment]
Hello all, VNET integration is now available in the [Premium plan](https://docs.microsoft.com/en-us/azure/azure-functions/functions-premium-plan#features). The Premium plan scales dynamically, and has a fixed cost of about 15% the initial cost of an ASE.

**Comment by mrtristan** [comment]
@alexkarcher-msft that's great. is doing the same for the consumption plan on the roadmap at all?

**Comment by alexkarcher-msft** [comment]
Not at the moment. The more modern VMs that the Premium plan runs on are what enable us to connect to a VNET while scaling dynamically.

**Comment by denniceritchie** [comment]
In premium plan and app service plan also the integration is not working.

**Comment by jeffhollan** [comment]
@denniceritchie can you potentially open another issue with details on what you are seeing in Premium / App Service Plan that isn't working?  I'm not aware of any issues that would prevent it it from working in the premium plan.

**Comment by Ratomir** [comment]
Does anyone have any updates regarding this issue? When we can expect some kind of solution in the preview, or we can completely dismiss expectations about this feature.

Spending 153$ per month for a Premium plan is a really huge amount of money for someone, let's say who has 3-5 functions.
![image](https://user-images.githubusercontent.com/8688494/90140942-bc1d3200-dd7a-11ea-8707-ef28ad5e1d5a.png)
Or I am doing wrong calculations. 

Ok, I know that I have an option to deploy my functions with Azure Container Instances and connect it with the VNET, but in that case, I have to worry about the time of running and other things, which is not quite a good solution. 

Kind regards

**Comment by forlayo** [comment]
@Ratomir I know that's not a "perfect" solution but you can use App Service Plan with S1 machines, that will allow you to use VNETs

**Comment by wpitallo** [comment]
Please, please make this available on consumption plan!

**Comment by 0xUnicorn** [comment]
Hi, I'm creating a startup making use of +10 function apps (consumption plan) for my microservice infrastructure and not being able to add them to same VNET's as their related services like StorageAccounts, keyvaults etc. really sucks.

Paying 152$ for each function app is not possible, and dumping all functions into the same function app isn't a proper solution.
A lot of my function apps are simple API's taking less than 1 second to execute and making use of different related services secured by vnet's.

So I miss being able to add VNET's to function apps running on consumption plan

**Comment by nzthiago** [comment]
> Hi, I'm creating a startup making use of +10 function apps (consumption plan) for my microservice infrastructure and not being able to add them to same VNET's as their related services like StorageAccounts, keyvaults etc. really sucks.
> 
> Paying 152$ for each function app is not possible, and dumping all functions into the same function app isn't a proper solution.
> A lot of my function apps are simple API's taking less than 1 second to execute and making use of different related services secured by vnet's.
> 
> So I miss being able to add VNET's to function apps running on consumption plan

Also agree that VNET integration in consumption would be great, but to clarify: you pay Premium for the app plan instance, and you can have up to 100 function apps in the same Premium Functions app plan. So you can still pay for one instance of the Premium plan and have your 10 function apps in the same plan.

**Comment by 0xUnicorn** [comment]
> > Hi, I'm creating a startup making use of +10 function apps (consumption plan) for my microservice infrastructure and not being able to add them to same VNET's as their related services like StorageAccounts, keyvaults etc. really sucks.
> > Paying 152$ for each function app is not possible, and dumping all functions into the same function app isn't a proper solution.
> > A lot of my function apps are simple API's taking less than 1 second to execute and making use of different related services secured by vnet's.
> > So I miss being able to add VNET's to function apps running on consumption plan
> 
> Also agree that VNET integration in consumption would be great, but to clarify: you pay Premium for the app plan instance, and you can have up to 100 function apps in the same Premium Functions app plan. So you can still pay for one instance of the Premium plan and have your 10 function apps in the same plan.

Oh Thanks so much @nzthiago , I didn't notice it is the App Plan you buy as Premium.

**Comment by wpitallo** [comment]
you can have up to 100 function apps in the same Premium Functions app plan :) :) :)

**Comment by ishepherd** [comment]
Where are we with getting Consumption Plan onto more modern VMs, so you can resolve this issue? 

Network-level access control is an essential, not a "premium" optional extra.

For example: there was this vuln in Cosmos DB where keys were leaked. https://www.theregister.com/2021/08/27/chaos_db_azure_cosmos_flaw/

**Comment by jbonfardeci** [comment]
Our only option is to synchronize data updates from our VM database cluster to an Azure Table in order to read the data we need into our Consumption Function apps. All our data is in a VM cluster under VNet. This is a major blocker and we do not need most of the features with Premium.

**Comment by ishepherd** [comment]
The hardware hosting Consumption Plan is old and doesn't support the 'SWIFT' networking tech (which allows packets to be teleported to/from a customer VNET)

I suspect: the cheap/free pricing of Consumption Plan is enabled solely by this hardware being otherwise redundant - fit only for retirement. Like, who wants instances that are slow and can't do secure networking properly?

I suspect: they have modelled what it would cost to run their Consumption model on modern hardware and it isn't profitable enough so it will never happen.

I suspect: this is why they pedalled back from per-invocation consumption pricing, when they designed the Premium Plan.

My 2c

**Comment by jbonfardeci** [comment]
I think you have a solid insight into the reasoning behind it. Had I known
this limitation months ago, our implementation may have been very different.

**Comment by Trubador** [comment]
Any news on this?

I am strongly considering moving to AWS because of VNET not being available in the consumption plan for Functions. This is an absolute must-have in terms of security and frankly, I am disappointed that Microsoft has not addressed this, since Azure is promoted to be highly secure.

**Comment by o-l-a-v** [comment]
For visibility it's probably a good idea to add votes to the new Azure Feedback page as well.

* https://feedback.azure.com/d365community/idea/d647e855-f224-ec11-b6e6-000d3a4f0da0

**Comment by ShanonJackson** [comment]
There's only 1 compute as a service "TRUE" serverless offering on Azure and its the function apps consumption plan, this issue is 4 years old now and there's still been no progress on making any kind of network security model available on consumption plan at all.

It's forcing your customers into a difficult decision as its either.
A: Pay more, over provision and swap onto a non-server-less based pricing tier
B: Open your 'X' up to the public internet, for which X will normally be a database.

Bigger companies may go for 'A' but for for SME's I have no doubt there will be a large number of them going for B which introduces significant security risks.

AWS has this VPC (ENI)
GCP has this. VPC access
Alibaba cloud has this. VPC (ENI)
Azure does not have this.

Please prioritize this feature so we can use FaaS in our company.

**Comment by RafaelRuales** [comment]
Is there any possible way that Azure could be more inferior than AWS? no VNET integration in consumption plans is like a balloon with no air in it.

**Comment by Trubador** [comment]
@ShanonJackson  I could not have said it better myself!

**Comment by Trubador** [comment]
> The hardware hosting Consumption Plan is old and doesn't support the 'SWIFT' networking tech (which allows packets to be teleported to/from a customer VNET)
> 
> I suspect: the cheap/free pricing of Consumption Plan is enabled solely by this hardware being otherwise redundant - fit only for retirement. Like, who wants instances that are slow and can't do secure networking properly?
> 
> I suspect: they have modelled what it would cost to run their Consumption model on modern hardware and it isn't profitable enough so it will never happen.
> 
> I suspect: this is why they pedalled back from per-invocation consumption pricing, when they designed the Premium Plan.
> 
> My 2c

I think you are right, but it does not make sense that AWS can solve this profitably and Microsoft has decided not to until now.

**Comment by PureKrome** [comment]
Could someone from the AzF team please help answer some of the above thoughts/guesses?

Also, is there a roadmap that has some answers targeting the above issues? Is there _any_ news at all, please?

**Comment by vigouredelaruse** [comment]
> > The hardware hosting Consumption Plan is old and doesn't support the 'SWIFT' networking tech (which allows packets to be teleported to/from a customer VNET)

what is your source for this pertinent information?

**Comment by Trubador** [comment]
> The hardware hosting Consumption Plan is old and doesn't support the 'SWIFT' networking tech (which allows packets to be teleported to/from a customer VNET)

@ishepherd mentioned this. I have no facts on this information, but I think it is probable.

**Comment by ishepherd** [comment]
@vigouredelaruse sorry I have no source on this. Just conversations with people in Azure 2-3 years ago.

**Comment by Slukad** [comment]
I can't believe that, after years of this issue being open, there's not a single answer if this feature is going to be included or not in the roadmap. This only shows lack of transparency on a very attractive service in theory but really useless in reality for production workloads since VNET integration is a must

**Comment by cachai2** [comment]
Hi everyone, thank you for the continued feedback. Agreed with a lot of the overall pain points folks have expressed on this thread. We're evaluating this from the product side.

**Comment by ursaciuc-adrian** [comment]
Any updates on this?

**Comment by alexferrari88** [comment]
what a bummer that Azure serverless doesn't have sensible security practices in place. I'm looking into migrating to AWS after discovering this 😔

**Comment by alejojarahi** [comment]
Any news or roadmap related to this? AWS has been have this functionality for years in Lambda +VPC and Azure today serverless is not viable.

Premium plan does not respect serverless principles

**Comment by o-l-a-v** [comment]
@alejojarahi 

Have you considered Azure Container Apps?

* https://endjin.com/blog/2022/09/bye-bye-azure-functions-hello-azure-container-apps-part-1-introduction

**Comment by dgonzalezpulido** [comment]
We need the integration of Consumption plan with VNET. Some customers doesnt use Consumption plan for its lack of secure conectivity or they doenst choose Azure directly. The vnet integration is needed!

**Comment by Trubador** [comment]
> Hi everyone, thank you for the continued feedback. Agreed with a lot of the overall pain points folks have expressed on this thread. We're evaluating this from the product side.

What is there to evaluate? This is the single most requested feature of Azure Functions by customers and the community and an absolute must in terms of security. What is blocking developing this feature?

**Comment by ursaciuc-adrian** [comment]
As far as I understood, the reason they can't offer vnet integration is that the hardware the functions on the consumption plan are using is old. It's so ancient that it can't support some new technologies needed for vnet communication.

**Comment by ursaciuc-adrian** [comment]
> @alejojarahi
> 
> Have you considered Azure Container Apps?
> 
> * https://endjin.com/blog/2022/09/bye-bye-azure-functions-hello-azure-container-apps-part-1-introduction

Switching to ACA is not as suitable as it might seem. It's quite difficult to do so for existing projects and it has its fair share of downsides.

**Comment by Trubador** [comment]
> As far as I understood, the reason they can't offer vnet integration is that the hardware the functions on the consumption plan are using is old. It's so ancient that it can't support some new technologies needed for vnet communication.

I think you are right on that. I guess technical debt is the single biggest reason why companies  have difficulties innovating.

**Comment by enzo-z** [comment]
Any updates on this? 5 years since this issue opened, and I am baffled that the Consumption Plan (true serveless) doesn't have a VNet Support...  Are there any workarounds that does not involve upgrading to Premium Plan? 😥

**Comment by Trubador** [comment]
@enzo-z I don't think this will ever be implemented. I have switched to Container Apps because of this.

**Comment by nzthiago** [comment]
Hi everyone - a little early to share specifics or timelines but we are working on this.

**Comment by Trubador** [comment]
> Hi everyone - a little early to share specifics or timelines, but we are working on this.

Awesome, thanks for giving us an update. If this gets implemented for the consumption function app plan, it will be a much welcome feature that will greatly enhance its usability. Can you share some information on the challenges that need to be overcome to implement this?

**Comment by nzthiago** [comment]
While there are some implementation changes for a 'scale to zero' plan to support VNet when compared to the other hosting plans, it's been mostly prioritization related so far. It's great to see continued interest in this feature!

**Comment by Trubador** [comment]
> While there are some implementation changes for a 'scale to zero' plan to support VNet when compared to the other hosting plans, it's been mostly prioritization related so far. It's great to see continued interest in this feature!

Thanks for the insights. I can not imagine that another feature than VNET support for function app running in consumption would be more requested by customers and the community, so I hope you will move this feature to the top of your prioritization.

**Comment by enzo-z** [comment]
> While there are some implementation changes for a 'scale to zero' plan to support VNet when compared to the other hosting plans, it's been mostly prioritization related so far. It's great to see continued interest in this feature!

Thank you very much for the attention, @nzthiago ! If the status of this feature does change - whether that new status is abandoned or started - please share updates with us. Hopefully all goes well and consumption users can have access to this integration xd

**Comment by DominicBoettger** [comment]
I still hope this feature will be available in the future

**Comment by fisica3** [comment]
This is very important for us, we have a lot of services deployed over the internal network, so connect to them in a Consumption model is required for some business case, is sad that sometimes we had to use another technology instead of Azure Functions.

**Comment by Ez2** [comment]
We (Rabobank) are also waiting for VNET integration for the Consumption plan.

**Comment by goldsperenso** [comment]
Same problem here. Consumption plans for functions don't support VNET integration and are a major blocker. Rinse & repeat the complaint(s) of all of the above since 2018.

**Comment by pihai** [comment]
It seems that something is in the pipeline: https://azure.microsoft.com/en-us/updates/register-for-access-to-upcoming-azure-functions-flex-consumption/

**Comment by gabbsmo** [comment]
Glad to hear that there is progress. Is Logic Apps also in the scope for "Flex Consumption Plan"? Currently you cannot limit incoming requests to a VNet to a HTTP triggered Logic App without "upgrading" to a plan with overprovisioning.

**Comment by kewear** [comment]
> Glad to hear that there is progress. Is Logic Apps also in the scope for "Flex Consumption Plan"? Currently you cannot limit incoming requests to a VNet to a HTTP triggered Logic App without "upgrading" to a plan with overprovisioning.

Hi @gabbsmo, I am a PM from Logic Apps team. Thanks for surfacing this question. Are you able to share more context about your scenario? Is this related to a single logic app workflow (similar to our current consumption sku) or related to our standard sku where we host multple workflows inside the context of an application?

We currently don't have this in our plan, but always interested in understanding how customers can better use our offering.

Thanks,

Kent

**Comment by kewear** [comment]
> Glad to hear that there is progress. Is Logic Apps also in the scope for "Flex Consumption Plan"? Currently you cannot limit incoming requests to a VNet to a HTTP triggered Logic App without "upgrading" to a plan with overprovisioning.

Hi @gabbsmo, I am a PM from Logic Apps team. Thanks for sufacing this question. Are you able to share more context about your scenario? Is this related to a single logic app workflow (similar to our current consumption sku) or related to our standard sku where we host multple workflows inside the context of an application?

**Comment by gabbsmo** [comment]
@kewear I guess neither. Standard SKU is not serverless. The current Logic Apps consumption plan does not allow for VNet integration.

What I am looking for is combining the cost benefits of serverless pay for what you use architecture combined with the security benefits of VNet integration. 

There is a new plan for Functions in the works that will offer this. Is there a plan for Logic Apps as well?

Typically the use case is thag you have a logic app interfacing with some backend service like SQL or Storage. To restrict access to these backend services you can restrict incoming traffic from a VNET.

**Comment by kewear** [comment]
@gabbsmo it is not currently in our plan, but this is something that we will solicit more feedback on from customers.

**Comment by ishepherd** [comment]
@kewear my 2c
* Microsoft should back up [all the fine words](https://azure.microsoft.com/en-us/explore/security/) with proactive action on cloud security. Don't wait till your customers demand action.
* Deprecate SKUs that don't support network security (VNETs)
* Help mitigate all this kind of stuff: https://www.cloudvulndb.org/results?tags=Azure

**Comment by StevenRasmussen** [comment]
Flex consumption is in preview in many areas: https://learn.microsoft.com/en-us/azure/azure-functions/flex-consumption-plan

**Comment by nzthiago** [comment]
Hi everyone - happy to announce that [Flex Consumption is now GA](https://aka.ms/flexconsumption/gaannouncement). Being a consumption-based plan with VNet capabilities, Flex Consumption addresses this issue. Going forward we will keep expanding Flex Consumption to more regions, close [gaps](https://learn.microsoft.com/en-us/azure/azure-functions/flex-consumption-plan#considerations), and work on migration documentation and features from Consumption to Flex Consumption. Thank you all for the input and patience while we addressed this key feedback, and the Azure Functions team can't wait to see what you build with Flex Consumption.
